teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+
+> :warning: The new API field is introduced as an `alpha` feature
+
+> :warning: This `enum` field should be specified if and only if the parameter type is `string`, which is similar to the existing [`properties`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) field that should be only specified for `object` type.
+
+### Tasks
+The following is a `Task` example with the `task-version` param bounded to a set of constants: `v1.21`, `v1.20` and `v1.19`: 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: golang-build
+spec:
+  params:
+  - name: task-version
+    description: golang version to use for builds
+    type: string
+    enum: # this field can be specified iff the type is string
+      - "v1.21"
+      - "v1.20"
+      - "v1.19"
+    default: "v1.21"
+  steps:
+  - name: build
+    image: docker.io/library/golang:$(params.task-version)
+    ...
+```
+
+#### TaskRuns with TaskRef or Remote Resolution
+`Task` users can provide a value for the parameter at the execution time. If the value is in the `enum` list specified by the author, the task will be executed normally. The task will fail the validation with reason `InvalidParamValue` otherwise. 
+
+As an example, the `TaskRun` references the above `Task`. The execution of the `TaskRun` should fail with reason `InvalidParamValue` and a detailed error message since `latest` is NOT in the `enum` list. 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  name: run-golang-build
+spec:
+  params:
+  - name: version
+    value: "latest" # 'latest' is not in the enum list
+  taskRef:
+    name: golang-build
+    ...
+```
+
+#### TaskRuns with Embedded TaskSpec
+`Task` users are allowed to specify `enum` for `TaskRuns` with Embedded `TaskSpec` and Tekton will perform the same validation as described above.
+
+``` yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  generateName: golang-build-run
+spec:
+  params:
+    - name: version
+      value: "v1.21"
+  taskSpec:
+    params:
+      - name: version
+        type: string 
+        enum:
+          - "v1.21"
+          - "v1.20"
+          - "v1.19"
+    ...
+```
+
+While we haven't identified any `enum` use case in this scenario at the time writing this TEP, this is treated as a valid `TaskRun` for `TaskSpec` API compatibility reason.
+
+### Pipelines
+`Pipeline` author can also define a `string` `param` with the `enum` keyword in the `PipelineSpec` `Param` and pass this `string` `param` to the referenced `Task`.
+
+> :warning: Tekton validates the **intersection** of enum set specified in a PipelineTask and the enum set specified in the Pipeline.
```

Comment:
Could you expand on this?
Does this mean the validation fails the `PipelineRun` upfront if the intersection is empty?

Reply:
Hi @afrittoli. Yeah, I have expanded the point below in the Pipeline using TaskRef or Embedded PipelineTask cases.

Yeah, it should fail the `PipelineTun` if the intersection is empty, but I think I need to make it more explicit that a Param without "Enum" refers to a set of all possible values (will update TEP to reflect it).

=> So a `Pipeline` cannot allow a param value that is not allowed by PipelineTask.

Reply:
Explained the scenario where enum is not specify in the param, PTAL!